File: sampleCode/R/utilities/rest.R
# The HTTP client every endpoint file uses.
#
# One place is responsible for everything that is the same on every call:
# attaching the bearer token, logging the exchange, turning an error response
# into an R condition you can read, backing off when the API asks you to, and
# refreshing an expired token once before giving up.
#
# Errors deserve a word. Every failure from this API comes back as a
# problem-details document (RFC 9457) with a `title`, a `detail`, and a
# `traceId`. Those three fields are the difference between "something went wrong"
# and knowing what to fix, so this client raises them as an `implan_api_error`
# condition rather than letting a failed call return NULL and break somewhere
# else with a confusing message.
#
# Most of the work is done by httr2's own policies rather than by hand:
#
#   req_auth_bearer_token()  attaches the token and redacts it when a request is
#                            printed
#   req_retry()              waits and tries again on 429 and the transient 5xx,
#                            honouring Retry-After when the API sends one
#   req_error()              puts the problem details into the R error message
#   req_throttle()           keeps the bulk workflows inside the published limits
#
# Wiki: Requests - https://github.com/Implan-Group/api/wiki/Requests
# Wiki: Responses - https://github.com/Implan-Group/api/wiki/Responses

library(httr2)
library(jsonlite)


# Statuses worth trying again. 429 means the rate limit was hit; the 5xx family
# here means the service was briefly unavailable rather than that the request was
# wrong.
RETRY_STATUSES <- c(429L, 502L, 503L, 504L)

MAX_ATTEMPTS <- 4L

# Sent on every request, so IMPLAN can tell sample traffic apart in their logs.
USER_AGENT <- "implan-impact-api-samples-r"


# The body the API returns with any error, per RFC 9457.
#
# `trace_id` is the one to quote when reporting a problem to IMPLAN support: it
# identifies the exact request in their logs.
problem_details <- function(resp) {
  status <- resp_status(resp)
  title <- resp_status_desc(resp)
  detail <- ""
  trace_id <- NULL

  body <- tryCatch(resp_body_string(resp), error = function(condition) "")

  parsed <- if (nzchar(trimws(body))) {
    tryCatch(fromJSON(body, simplifyVector = FALSE), error = function(condition) NULL)
  } else {
    NULL
  }

  if (is.list(parsed)) {
    if (!is.null(parsed$title)) title <- parsed$title
    if (!is.null(parsed$detail)) detail <- parsed$detail
    if (!is.null(parsed$traceId)) trace_id <- parsed$traceId
  } else {
    # Not JSON at all: a gateway error page, or an empty body. Use whatever text
    # came back so the reader is not left with nothing.
    detail <- substr(trimws(body), 1, 1000)
  }

  structure(
    list(status = status, title = title, detail = detail, trace_id = trace_id),
    class = "implan_problem"
  )
}


# One readable block, for a console message or a support ticket.
describe.implan_problem <- function(x, ...) {
  lines <- sprintf("%d %s", x$status, x$title)
  if (nzchar(x$detail)) {
    lines <- c(lines, paste0("  ", x$detail))
  }
  if (!is.null(x$trace_id)) {
    lines <- c(lines, paste0("  traceId: ", x$trace_id))
  }
  paste(lines, collapse = "\n")
}


# Raised when the API answers with an error.
#
# Catch it with tryCatch(..., implan_api_error = function(cnd) ...) to handle an
# expected failure, for example a 409 when a title is already taken.
# `cnd$status` is the status code and `cnd$problem$detail` is the API's
# explanation.
implan_api_error <- function(method, url, problem) {
  structure(
    class = c("implan_api_error", "error", "condition"),
    list(
      message = paste0(method, " ", url, "\n", describe(problem)),
      call = NULL,
      problem = problem,
      status = problem$status,
      url = url
    )
  )
}


# Raised when the request never reached the API at all.
implan_transport_error <- function(base_url, condition) {
  structure(
    class = c("implan_transport_error", "error", "condition"),
    list(
      message = paste0(
        "Could not reach ", base_url, ".\n", conditionMessage(condition), "\n",
        "Check your network connection and IMPLAN_API_URL."
      ),
      call = NULL
    )
  )
}


# Builds the client the workflows pass around.
#
# An environment rather than a list, because the token and the rate limit are
# mutable: a token refreshed part way through a run has to be visible to every
# call that follows, wherever the client was passed to.
api_client <- function(token_provider,
                       token_refresher = NULL,
                       base_url = implan_base_url(),
                       requests_per_minute = NULL) {
  client <- new.env(parent = emptyenv())
  client$base_url <- sub("/+$", "", base_url)
  client$token_provider <- token_provider
  client$token_refresher <- token_refresher
  client$requests_per_minute <- requests_per_minute
  class(client) <- "implan_client"
  client
}


# Switches on client-side throttling for a bulk workflow.
set_rate_limit <- function(client, requests_per_minute) {
  client$requests_per_minute <- requests_per_minute
  invisible(client)
}


# Strips the `Bearer ` prefix, because req_auth_bearer_token() adds its own.
bare_token <- function(token) sub("^Bearer\\s+", "", token)


# Builds one request, with every policy this API needs attached.
#
# `query` is a named list. A value holding several elements becomes a repeated
# parameter (`regions=Oregon&regions=Wisconsin`), which is the shape this API
# expects for its filters, hence `.multi = "explode"`. A list would flatten to
# the last value alone, silently, which is why the endpoint files build plain
# character vectors here.
#
# `body` is sent as JSON. It is allowed on a GET, which sounds wrong but is what
# a few of this API's report endpoints require; see
# `get_estimated_growth_percentage()` in endpoints/impact_results.R. httr2
# switches the method to POST as soon as a body is attached, so the method is set
# again afterwards.
build_request <- function(client, path, method = "GET", query = NULL, body = NULL) {
  req <- request(paste0(client$base_url, path)) |>
    req_user_agent(USER_AGENT) |>
    req_timeout(REQUEST_TIMEOUT_SECONDS) |>
    req_auth_bearer_token(bare_token(client$token_provider())) |>
    req_retry(
      max_tries = MAX_ATTEMPTS,
      retry_on_failure = TRUE,
      is_transient = function(resp) resp_status(resp) %in% RETRY_STATUSES
    ) |>
    req_error(body = function(resp) describe(problem_details(resp)))

  query <- Filter(Negate(is.null), as.list(query))
  if (length(query) > 0) {
    req <- req_url_query(req, !!!query, .multi = "explode")
  }

  if (!is.null(body)) {
    req <- req_body_json(req, body, auto_unbox = TRUE, null = "null")
  }

  if (!is.null(client$requests_per_minute)) {
    # The published limits are per minute, so one bucket that fills over 60
    # seconds is exactly the right shape.
    req <- req_throttle(req, capacity = client$requests_per_minute, fill_time_s = 60)
  }

  req_method(req, method)
}


# Performs one request, refreshing the token once if it has expired.
#
# httr2's retry policy has already dealt with 429 and the transient 5xx by the
# time anything reaches here, so this only has to handle the 401 case and turn
# whatever is left into an `implan_api_error`.
perform_request <- function(client, req) {
  for (attempt in seq_len(2L)) {
    started <- Sys.time()

    outcome <- tryCatch(
      list(ok = TRUE, resp = req_perform(req)),
      httr2_http = function(condition) {
        failed <- condition$resp
        if (is.null(failed)) {
          failed <- last_response()
        }
        list(ok = FALSE, resp = failed)
      },
      httr2_failure = function(condition) {
        stop(implan_transport_error(client$base_url, condition))
      }
    )

    log_exchange(req, outcome$resp, as.numeric(difftime(Sys.time(), started, units = "secs")))

    if (isTRUE(outcome$ok)) {
      return(outcome$resp)
    }

    # An expired token looks like a 401. Refresh once and try again; if the
    # second attempt also fails, the credentials or the subscription are the
    # problem, not the token.
    can_refresh <- attempt == 1L &&
      resp_status(outcome$resp) == 401L &&
      !is.null(client$token_refresher)

    if (can_refresh) {
      log_debug("401 received; refreshing the bearer token and retrying")
      client$token_refresher()
      req <- req_auth_bearer_token(req, bare_token(client$token_provider()))
      next
    }

    stop(implan_api_error(request_method(req), request_url(req), problem_details(outcome$resp)))
  }

  stop("The request loop finished without a response, which should not happen.", call. = FALSE)
}


# GETs a JSON document and parses it.
#
# `simplifyVector = FALSE` keeps every response as nested lists. It costs a
# little convenience and buys predictability: jsonlite would otherwise turn an
# array of objects into a data frame whenever those objects happen to share
# their fields, and leave it as a list when they do not, so the same endpoint
# would hand back different shapes on different days.
get_json <- function(client, path, query = NULL, body = NULL) {
  parse_json_response(perform_request(client, build_request(client, path, "GET", query, body)))
}


# GETs a text document, which for this API means a CSV report.
#
# Write the result to a file with a `.csv` extension and it opens in Excel or
# Sheets.
get_text <- function(client, path, query = NULL, body = NULL) {
  resp <- perform_request(client, build_request(client, path, "GET", query, body))
  resp_body_string(resp)
}


# POSTs a JSON body and reads whatever comes back.
#
# A few endpoints answer with a bare number or a sentence rather than an object:
# RunImpact returns a run id, and CancelImpact returns a confirmation. Both still
# need reading, so an unparseable response comes back as trimmed text.
post_json <- function(client, path, body = NULL) {
  parse_json_response(perform_request(client, build_request(client, path, "POST", body = body)))
}


# POSTs a file as multipart form data, for the Event Template upload.
post_file <- function(client, path, field_name, file_path) {
  req <- request(paste0(client$base_url, path)) |>
    req_user_agent(USER_AGENT) |>
    req_timeout(REQUEST_TIMEOUT_SECONDS) |>
    req_auth_bearer_token(bare_token(client$token_provider())) |>
    req_error(body = function(resp) describe(problem_details(resp)))

  # The field name matters: the API reads the upload from `excelFile`.
  parts <- list(curl::form_file(file_path))
  names(parts) <- field_name
  req <- do.call(req_body_multipart, c(list(req), parts))

  parse_json_response(perform_request(client, req))
}


# PUTs a JSON body and reads whatever comes back.
put_json <- function(client, path, body = NULL) {
  parse_json_response(perform_request(client, build_request(client, path, "PUT", body = body)))
}


# DELETEs a resource. The API returns no body worth reading.
delete_resource <- function(client, path) {
  perform_request(client, build_request(client, path, "DELETE"))
  invisible(NULL)
}


# Returns the response as parsed JSON when it is JSON, and as trimmed text
# otherwise.
parse_json_response <- function(resp) {
  body <- tryCatch(resp_body_string(resp), error = function(condition) "")
  if (!nzchar(trimws(body))) {
    return(NULL)
  }

  parsed <- tryCatch(fromJSON(body, simplifyVector = FALSE), error = function(condition) NULL)
  if (is.null(parsed)) {
    return(gsub('^"|"$', "", trimws(body)))
  }
  parsed
}

File: sampleCode/Python/utilities/rest.py
"""The HTTP client every endpoint module uses.

One class, `ApiClient`, is responsible for everything that is the same on every
call: attaching the bearer token, logging the exchange, turning an error response
into a Python exception you can read, backing off when the API asks you to, and
refreshing an expired token once before giving up.

Errors deserve a word. Every failure from this API comes back as a problem-details
document (RFC 9457) with a `title`, a `detail`, and a `traceId`. Those three fields
are the difference between "something went wrong" and knowing what to fix, so this
client raises them as an `ImplanApiError` rather than letting a failed call return
`None` and break somewhere else with a confusing message.

Wiki: Requests - https://github.com/Implan-Group/api/wiki/Requests
Wiki: Responses - https://github.com/Implan-Group/api/wiki/Responses
"""

import dataclasses
import datetime
import json
import random
import time
from collections import deque
from typing import Any, Callable, Mapping, Sequence

import requests

from utilities import config, logging_setup

# Statuses worth trying again. 429 means the rate limit was hit; the 5xx family
# here means the service was briefly unavailable rather than that the request was
# wrong. A 503 from the auth endpoint in particular is usually transient.
_RETRY_STATUSES = frozenset({429, 502, 503, 504})

_MAX_ATTEMPTS = 4


@dataclasses.dataclass(frozen=True)
class ProblemDetails:
    """The body the API returns with any error, per RFC 9457.

    `trace_id` is the one to quote when reporting a problem to IMPLAN support: it
    identifies the exact request in their logs.
    """

    status: int
    title: str
    detail: str
    trace_id: str | None = None
    type: str | None = None
    instance: str | None = None

    @classmethod
    def from_response(cls, response: requests.Response) -> "ProblemDetails":
        """Read problem details out of a response, whatever shape it arrived in."""
        title = response.reason or "Request failed"
        detail = ""
        trace_id = None
        problem_type = None
        instance = None

        try:
            body = response.json()
        except ValueError:
            # Not JSON at all: a gateway error page, or an empty body. Use whatever
            # text came back so the reader is not left with nothing.
            detail = (response.text or "").strip()[:1000]
        else:
            if isinstance(body, dict):
                title = body.get("title") or title
                detail = body.get("detail") or ""
                trace_id = body.get("traceId") or body.get("trace_id")
                problem_type = body.get("type")
                instance = body.get("instance")
            else:
                detail = str(body)[:1000]

        return cls(
            status=response.status_code,
            title=title,
            detail=detail,
            trace_id=trace_id,
            type=problem_type,
            instance=instance,
        )

    def describe(self) -> str:
        """One readable block, suitable for a console message or a support ticket."""
        lines = [f"{self.status} {self.title}"]
        if self.detail:
            lines.append(f"  {self.detail}")
        if self.trace_id:
            lines.append(f"  traceId: {self.trace_id}")
        return "\n".join(lines)


class ImplanApiError(RuntimeError):
    """Raised when the API answers with an error.

    Catch this to handle an expected failure, for example a 409 when a title is
    already taken. `error.problem.status` is the status code and
    `error.problem.detail` is the API's explanation.
    """

    def __init__(self, method: str, url: str, problem: ProblemDetails) -> None:
        super().__init__(f"{method} {url}\n{problem.describe()}")
        self.method = method
        self.url = url
        self.problem = problem

    @property
    def status_code(self) -> int:
        return self.problem.status


class RateLimiter:
    """Keeps a category of requests under a published per-minute limit.

    The API publishes limits per group of endpoints, and exceeding one earns a
    `429 Too Many Requests` and, if you keep going, a temporary ban. The bulk
    workflows switch this on so that a loop over thousands of regions stays
    polite; the short workflows do not need it.

    Limits are on the wiki home page:
    https://github.com/Implan-Group/api/wiki
    """

    def __init__(self, requests_per_minute: int) -> None:
        self.requests_per_minute = requests_per_minute
        self._timestamps: deque[float] = deque()

    def wait(self) -> None:
        """Block until another request would be within the limit."""
        now = time.monotonic()
        # Forget anything that happened more than a minute ago.
        while self._timestamps and now - self._timestamps[0] >= 60.0:
            self._timestamps.popleft()

        if len(self._timestamps) >= self.requests_per_minute:
            sleep_for = 60.0 - (now - self._timestamps[0]) + 0.1
            if sleep_for > 0:
                logging_setup.get_logger().debug(
                    "Rate limit: waiting %.1fs to stay under %d requests per minute",
                    sleep_for,
                    self.requests_per_minute,
                )
                time.sleep(sleep_for)

        self._timestamps.append(time.monotonic())


class ApiClient:
    """Sends authenticated requests to the Impact API.

    Build one of these with `utilities.auth.create_client()` rather than directly,
    so the token cache is wired up for you.
    """

    def __init__(
        self,
        token_provider: Callable[[], str],
        token_refresher: Callable[[], str] | None = None,
        base_url: str = config.BASE_URL,
    ) -> None:
        self._token_provider = token_provider
        self._token_refresher = token_refresher
        self.base_url = base_url.rstrip("/")
        self._session = requests.Session()
        self._logger = logging_setup.get_logger()
        # Set by the bulk workflows; `None` means no client-side throttling.
        self.rate_limiter: RateLimiter | None = None

    # -- the request that every other method goes through --------------------

    def _send(
        self,
        method: str,
        path: str,
        *,
        params: Mapping[str, Any] | None = None,
        json_body: Any | None = None,
        files: Mapping[str, Any] | None = None,
        expected_statuses: Sequence[int] = (200, 201, 202, 204),
    ) -> requests.Response:
        """Send one request, retrying transient failures, and raise on an error.

        `json_body` is serialized and sent as the request body. It is allowed on a
        GET, which sounds wrong but is what a few of this API's report endpoints
        require; see `endpoints.impact_results.get_estimated_growth_percentage`.
        """
        url = path if path.startswith("http") else f"{self.base_url}{path}"
        body = json.dumps(json_body) if json_body is not None else None

        last_problem: ProblemDetails | None = None

        for attempt in range(1, _MAX_ATTEMPTS + 1):
            if self.rate_limiter is not None:
                self.rate_limiter.wait()

            headers = {"Authorization": self._token_provider()}
            if body is not None:
                headers["Content-Type"] = "application/json"

            request = requests.Request(
                method=method,
                url=url,
                headers=headers,
                params=params,
                data=body,
                files=files,
            )
            prepared = self._session.prepare_request(request)

            started = datetime.datetime.now()
            response = self._session.send(
                prepared, timeout=config.REQUEST_TIMEOUT_SECONDS
            )
            elapsed = datetime.datetime.now() - started

            logging_setup.log_exchange(prepared, response, elapsed)

            if response.status_code in expected_statuses:
                return response

            problem = ProblemDetails.from_response(response)
            last_problem = problem

            # An expired token looks like a 401. Refresh once and try again; if the
            # second attempt also fails, the credentials or the subscription are
            # the problem, not the token.
            if (
                response.status_code == 401
                and self._token_refresher is not None
                and attempt == 1
            ):
                self._logger.debug("401 received; refreshing the bearer token and retrying")
                self._token_refresher()
                continue

            if response.status_code in _RETRY_STATUSES and attempt < _MAX_ATTEMPTS:
                delay = self._retry_delay(response, attempt)
                self._logger.info(
                    "  %s from the API; waiting %.0fs and trying again (attempt %d of %d)",
                    response.status_code,
                    delay,
                    attempt + 1,
                    _MAX_ATTEMPTS,
                )
                time.sleep(delay)
                continue

            raise ImplanApiError(method, url, problem)

        # Every attempt was a retryable failure.
        raise ImplanApiError(method, url, last_problem)  # type: ignore[arg-type]

    @staticmethod
    def _retry_delay(response: requests.Response, attempt: int) -> float:
        """How long to wait before trying again.

        The API's `Retry-After` header wins when it is present. Otherwise back off
        exponentially with a little randomness, so that a batch of parallel scripts
        does not retry in lockstep.
        """
        retry_after = response.headers.get("Retry-After")
        if retry_after:
            try:
                return float(retry_after)
            except ValueError:
                pass
        return min(60.0, (2.0**attempt) + random.uniform(0, 1))

    # -- typed helpers the endpoint modules call -----------------------------

    def get_json(
        self,
        path: str,
        *,
        params: Mapping[str, Any] | None = None,
        json_body: Any | None = None,
    ) -> Any:
        """GET a JSON document."""
        return self._send("GET", path, params=params, json_body=json_body).json()

    def get_text(
        self,
        path: str,
        *,
        params: Mapping[str, Any] | None = None,
        json_body: Any | None = None,
    ) -> str:
        """GET a text document, which for this API means a CSV report.

        The response is decoded as UTF-8 text. Write it to a file with a `.csv`
        extension and it opens in Excel or Sheets.
        """
        response = self._send("GET", path, params=params, json_body=json_body)
        # `requests` guesses an encoding from the headers and guesses badly for
        # `text/csv`; the API sends UTF-8, so say so before reading `.text`.
        response.encoding = response.encoding or "utf-8"
        return response.text

    def get_bytes(
        self, path: str, *, params: Mapping[str, Any] | None = None
    ) -> bytes:
        """GET a binary document, such as a zipped export."""
        return self._send("GET", path, params=params).content

    def post_json(
        self,
        path: str,
        *,
        json_body: Any | None = None,
        params: Mapping[str, Any] | None = None,
    ) -> Any:
        """POST a JSON body and read a JSON response.

        A few endpoints answer with a bare number or an empty body rather than an
        object, so an unparseable response comes back as raw text instead of
        raising.
        """
        response = self._send("POST", path, params=params, json_body=json_body)
        return _json_or_text(response)

    def post_file(self, path: str, *, field_name: str, file_path) -> Any:
        """POST a file as multipart form data, for the Event Template upload."""
        with open(file_path, "rb") as handle:
            files = {field_name: (file_path.name, handle)}
            response = self._send("POST", path, files=files)
        return _json_or_text(response)

    def put_json(self, path: str, *, json_body: Any | None = None) -> Any:
        """PUT a JSON body and read whatever comes back."""
        return _json_or_text(self._send("PUT", path, json_body=json_body))

    def patch_json(self, path: str, *, json_body: Any | None = None) -> Any:
        """PATCH a JSON body and read whatever comes back."""
        return _json_or_text(self._send("PATCH", path, json_body=json_body))

    def delete(self, path: str) -> None:
        """DELETE a resource. The API returns no body worth reading."""
        self._send("DELETE", path)


def _json_or_text(response: requests.Response) -> Any:
    """Return the response as JSON when it is JSON, and as trimmed text otherwise.

    `POST /api/v1/impact/{projectId}` answers with a bare run id, and
    `PUT /api/v1/impact/cancel/{runId}` answers with a sentence, so neither is a
    JSON object. Both still need reading.
    """
    if not response.content:
        return None
    try:
        return response.json()
    except ValueError:
        return response.text.strip().strip('"')
